### PR TITLE
feat: compiler API method for enriching error [APE-808]

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

fixes: #690 
Fixes: APE-193

Give compilers the chance to create a custom exception.

Will be used:

* For vyper compiler errors, like math errors
* For solidity custom error abi types

### How I did it

new api method
passthrough for not implemented.

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by unparalleled-js** [comment]
- [ ] Add JSON compiler implementation to make use of error ABIs.

File: src/ape_pm/compiler.py [hunk 1]
```diff
@@ -78,3 +80,62 @@ def compile(
             contract_types.append(contract_type)
 
         return contract_types
+
+    def enrich_error(self, err: ContractLogicError) -> ContractLogicError:
```
Comment (unparalleled-js) [inline src/ape_pm/compiler.py:84]: one thing is that this is exactly the same implementation `ape-solidity` would have... wondering if we should use a shared method somehow?

Comment (unparalleled-js) [reply src/ape_pm/compiler.py:84]: - [ ] going to make ecosystem decode_error method, that should help

**Comment by unparalleled-js** [comment]
To see it work in Solidity, check out this PR: https://github.com/ApeWorX/ape-solidity/pull/93

Example uncaught exception:

```
  File "/Users/jules/ApeProjects/ape-playground/scripts/haserror.py", line 8, in main
    contract.withdraw(sender=other)

ERROR: (Unauthorized) addr=0xc89D42189f0450C2b2c3c61f58Ec5d628176A1E7, counter=123
```

**Comment by unparalleled-js** [comment]
Can do [1223](https://github.com/ApeWorX/ape/issues/1223) right after

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**